### PR TITLE
Feature send and print headers of messages

### PR DESCRIPTION
**Pull request by schadin**
Hello!

I have made a small improvement in your wonderful client.
We have a need to send headers along with the messages. So I improved the parser and producer, now in the editor you can specify a comma-separated list of headers in format key=value.
image

It would also be nice to view these headers in the consumer, so I modified the output of the consumer 
image
The output of headers is rendered by setting kafka.consumers.headers.print, the default value is false.
image

### CODE REVIEW

**Comment by angelozerr** [comment]
I have just read quickly your code and it looks like promising. Could you update the doc https://github.com/jlandersen/vscode-kafka/blob/master/docs/Producing.md with your parameter in kafka file and new settings please.

Thank you very much for your contribution.

**Comment by angelozerr** [comment]
Please add in CHANGELOG.md a link with your issue

File: src/client/producer.ts [hunk 1]
```diff
@@ -185,6 +185,7 @@ export interface ProducerInfoUri {
     topicId?: string;
     key?: string;
     value?: string;
+    headers: Map<string, string>;
```
Comment (angelozerr) [inline src/client/producer.ts:188]: It should be headers?: ....

File: src/commands/producers.ts [hunk 2]
```diff
@@ -57,6 +57,12 @@ export class ProduceRecordCommandHandler {
             }
 
             const messages = [...Array(times).keys()].map(() => {
+
+                let messageHeaders: IHeaders = {}
```
Comment (angelozerr) [inline src/commands/producers.ts:61]: Use const

File: src/client/producer.ts (hunk 1 above)
Comment (schadin) [reply src/client/producer.ts]: fix it

File: src/commands/producers.ts (hunk 2 above)
Comment (schadin) [reply src/commands/producers.ts]: fix it

**Comment by schadin** [comment]
> I have just read quickly your code and it looks like promising. Could you update the doc https://github.com/jlandersen/vscode-kafka/blob/master/docs/Producing.md with your parameter in kafka file and new settings please.
> 
> Thank you very much for your contribution.

@angelozerr Documentation and CHANGELOG was updated.

File: package.json [hunk 3]
```diff
@@ -74,6 +74,11 @@
                     "default": "latest",
                     "markdownDescription": "Initial offset to read from for new consumers. See [here](command:vscode-kafka.open.docs.page?%5B%7B%22page%22%3A%22Consuming%22%2C%22section%22%3A%22kafkaconsumersoffset%22%7D%5D) for more information."
                 },
+                "kafka.consumers.headers.print" : {
+                    "type": "boolean",
+                    "default": "false",
+                    "markdownDescription": "Print headers of message"
```
Comment (angelozerr) [inline package.json:80]: Add a link to your doc with `See [here](command:vscode-kafka.open.docs.page?...`

File: src/kafka-file/languageservice/services/codeLensProvider.ts [hunk 4]
```diff
@@ -114,6 +114,7 @@ export class KafkaFileCodeLenses {
         let keyFormatSettings: Array<SerializationSetting> | undefined;
         let valueFormat;
         let valueFormatSettings: Array<SerializationSetting> | undefined;
+        let headers: Map<String, String> = new Map();
```
Comment (angelozerr) [inline src/kafka-file/languageservice/services/codeLensProvider.ts:117]: To be consistent with other values, you not create a new Map.

`let headers: Map<String, String> | undefined;`

File: src/kafka-file/languageservice/services/codeLensProvider.ts [hunk 5]
```diff
@@ -134,6 +135,12 @@ export class KafkaFileCodeLenses {
                     valueFormatSettings = this.getSerializationSettings(callee);
                     break;
                 }
+                case 'headers': {
+                    if (property.propertyValue) {
+                        this.parseHeaders(property.propertyValue, headers);
```
Comment (angelozerr) [inline src/kafka-file/languageservice/services/codeLensProvider.ts:140]: Update parseHeader to write:

`headers = this.parseHeaders(property.propertyValue);`

File: src/kafka-file/languageservice/services/codeLensProvider.ts [hunk 6]
```diff
@@ -144,10 +151,11 @@ export class KafkaFileCodeLenses {
             messageKeyFormat: keyFormat,
             messageKeyFormatSettings: keyFormatSettings,
             messageValueFormat: valueFormat,
-            messageValueFormatSettings: valueFormatSettings
+            messageValueFormatSettings: valueFormatSettings,
+            headers
         } as ProduceRecordCommand;
     }
-
+
```
Comment (angelozerr) [inline src/kafka-file/languageservice/services/codeLensProvider.ts:158]: remove extra spaces

File: docs/Producing.md [hunk 7]
```diff
@@ -110,6 +112,7 @@ For instance:
 
 PRODUCER keyed-message
 topic: fakerjs-topic
+headers: messageId={{random.uuid}}
```
Comment (angelozerr) [inline docs/Producing.md:115]: could you add a new hearder parameter to see that comma isused please.

Comment (schadin) [reply docs/Producing.md]: Added second header

File: src/kafka-file/languageservice/services/codeLensProvider.ts (hunk 6 above)
Comment (schadin) [reply src/kafka-file/languageservice/services/codeLensProvider.ts]: Oops... sorry, fixed

File: src/kafka-file/languageservice/services/codeLensProvider.ts (hunk 5 above)
Comment (schadin) [reply src/kafka-file/languageservice/services/codeLensProvider.ts]: fixed

File: src/kafka-file/languageservice/services/codeLensProvider.ts (hunk 4 above)
Comment (schadin) [reply src/kafka-file/languageservice/services/codeLensProvider.ts]: fixed

File: package.json (hunk 3 above)
Comment (schadin) [reply package.json]: Link added

**Comment by angelozerr** [comment]
I think it should be nice to write some tests in https://github.com/jlandersen/vscode-kafka/blob/86f0b7326d8f2c67c7eaba2a69861b4b4e6252cb/src/test/suite/kafka-file/languageservice/parser.test.ts#L105 with hearders content like:

 * heanders: a
 * heanders: a=b
 * heanders: a=b,
 * heanders: a=b,c
 * heanders: a=b,c=d

File: src/commands/producers.ts [hunk 8]
```diff
@@ -66,16 +72,21 @@ export class ProduceRecordCommandHandler {
                     const randomizedKey = (key) ? faker.fake(key) : key;
                     faker.seed(seed);
                     const randomizedValue = faker.fake(value);
+                    if (headers && headers.size > 0) {
```
Comment (fbricon) [inline src/commands/producers.ts:75]: if you want to use the same seed for headers, as you documented, then you need to:
```json
Object.keys(messageHeaders).forEach(val => {
   faker.seed(seed);
   messageHeaders[val] = faker.fake(messageHeaders[val] as string;
});

```

**Comment by angelozerr** [comment]
@schadin is there any chance that you could write tests like I suggested. 

I think we will do a release ASAP when https://github.com/jlandersen/vscode-kafka/issues/211 will be fixed. If this release could host your great work, it should b ereally nice. Thanks!

**Comment by schadin** [comment]
> @schadin is there any chance that you could write tests like I suggested.
> 
> I think we will do a release ASAP when #211 will be fixed. If this release could host your great work, it should b ereally nice. Thanks!

@angelozerr Hi! I try in this week. Hard deadline on work.

**Comment by angelozerr** [comment]
> @angelozerr Hi! I try in this week. Hard deadline on work.

No rush please! Do that when you will have time.

**Comment by schadin** [comment]
> > @angelozerr Hi! I try in this week. Hard deadline on work.
> 
> No rush please! Do that when you will have time.

@angelozerr Hi! I'm wrote one test. It right?
Parser return two element: key "header" and value "value after header". 
If you think the parser should determine the token types, then I need to rewrite the parser, because now the headers and values to send are received just before it.

Good day!

**Comment by angelozerr** [comment]
Thanks so much @schadin for your great contribution!

**Pull request merged**